File: src/KeyStats.Core/KeyboardCounter.cs
namespace KeyStats.Core;

public sealed class KeyboardCounter
{
    private readonly long[] _counts = new long[KeyCatalog.Count];
    private readonly HashSet<PressedKey> _pressed = [];
    private readonly object _pressedGate = new();
    private long _totalCount;
    private int _isPaused;

    public bool IsPaused => Volatile.Read(ref _isPaused) != 0;

    public bool Process(nint deviceHandle, RawKeyEvent keyEvent)
        => ProcessDetailed(deviceHandle, keyEvent).Recognized;

    public KeyProcessResult ProcessDetailed(nint deviceHandle, RawKeyEvent keyEvent)
    {
        var normalized = KeyNormalizer.Normalize(keyEvent);
        if (normalized is not { } keyId)
        {
            return KeyProcessResult.Unrecognized;
        }

        return ProcessButton(deviceHandle, keyId, keyEvent.IsBreak);
    }

    public KeyProcessResult ProcessButton(nint deviceHandle, KeyId keyId, bool isBreak)
    {
        if ((uint)keyId >= (uint)KeyCatalog.Count)
        {
            return KeyProcessResult.Unrecognized;
        }

        // Tab 在部分 Windows 10 环境的 Alt+Tab 系统切换中不会产生 Raw Input，
        // 因而需要与补偿采样共用一个全局身份，避免正常消息和补偿消息重复计数。
        var trackingDeviceHandle = keyId == KeyId.Tab ? 0 : deviceHandle;
        var pressedKey = new PressedKey(trackingDeviceHandle, keyId);
        if (isBreak)
        {
            lock (_pressedGate)
            {
                _pressed.Remove(pressedKey);
            }

            return KeyProcessResult.RecognizedOnly(keyId);
        }

        bool isNewPress;
        lock (_pressedGate)
        {
            isNewPress = _pressed.Add(pressedKey);
        }

        if (isNewPress && !IsPaused)
        {
            Interlocked.Increment(ref _counts[(int)keyId]);
            Interlocked.Increment(ref _totalCount);
            return KeyProcessResult.CountedPress(keyId);
        }

        return KeyProcessResult.RecognizedOnly(keyId);
    }

    public void SetPaused(bool paused) => Interlocked.Exchange(ref _isPaused, paused ? 1 : 0);

    public void ClearCounts(bool resetPressedState = false)
    {
        foreach (ref var count in _counts.AsSpan())
        {
            Interlocked.Exchange(ref count, 0);
        }

        Interlocked.Exchange(ref _totalCount, 0);

        if (resetPressedState)
        {
            lock (_pressedGate)
            {
                _pressed.Clear();
            }
        }
    }

    public void RemoveDevice(nint deviceHandle)
    {
        lock (_pressedGate)
        {
            _pressed.RemoveWhere(key => key.DeviceHandle == deviceHandle);
        }
    }

    public void ReleaseKey(KeyId keyId)
    {
        lock (_pressedGate)
        {
            _pressed.RemoveWhere(key => key.KeyId == keyId);
        }
    }

    public KeyboardSnapshot GetSnapshot()
    {
        var counts = new long[_counts.Length];
        for (var index = 0; index < _counts.Length; index++)
        {
            counts[index] = Interlocked.Read(ref _counts[index]);
        }

        HashSet<KeyId> pressedKeys;
        lock (_pressedGate)
        {
            pressedKeys = _pressed.Select(key => key.KeyId).ToHashSet();
        }

        return new KeyboardSnapshot(
            Interlocked.Read(ref _totalCount),
            counts,
            pressedKeys,
            IsPaused);
    }

    private readonly record struct PressedKey(nint DeviceHandle, KeyId KeyId);
}

File: src/core/keyboard_counter.cpp
#include "keyboard_counter.hpp"

#include "key_normalizer.hpp"

#include <algorithm>
#include <atomic>

namespace keystats {

bool KeyboardCounter::IsPaused() const noexcept {
    return std::atomic_ref<const int>(is_paused_).load(std::memory_order_acquire) != 0;
}

bool KeyboardCounter::Process(std::intptr_t device_handle, const RawKeyEvent& key_event) {
    return ProcessDetailed(device_handle, key_event).recognized;
}

KeyProcessResult KeyboardCounter::ProcessDetailed(std::intptr_t device_handle, const RawKeyEvent& key_event) {
    const auto normalized = KeyNormalizer::Normalize(key_event);
    if (!normalized) {
        return KeyProcessResult::Unrecognized();
    }
    return ProcessButton(device_handle, *normalized, key_event.IsBreak());
}

KeyProcessResult KeyboardCounter::ProcessButton(std::intptr_t device_handle, KeyId key_id, bool is_break) {
    if (static_cast<unsigned>(ToIndex(key_id)) >= static_cast<unsigned>(kKeyCount)) {
        return KeyProcessResult::Unrecognized();
    }

    const auto tracking_device = key_id == KeyId::Tab ? 0 : device_handle;
    const PressedKey pressed_key{tracking_device, key_id};
    if (is_break) {
        std::lock_guard lock(pressed_gate_);
        pressed_.erase(pressed_key);
        return KeyProcessResult::RecognizedOnly(key_id);
    }

    bool is_new_press = false;
    {
        std::lock_guard lock(pressed_gate_);
        is_new_press = pressed_.insert(pressed_key).second;
    }

    if (is_new_press && !IsPaused()) {
        std::atomic_ref(counts_[static_cast<std::size_t>(ToIndex(key_id))]).fetch_add(1, std::memory_order_relaxed);
        std::atomic_ref(total_count_).fetch_add(1, std::memory_order_relaxed);
        return KeyProcessResult::CountedPress(key_id);
    }
    return KeyProcessResult::RecognizedOnly(key_id);
}

void KeyboardCounter::SetPaused(bool paused) noexcept {
    std::atomic_ref(is_paused_).store(paused ? 1 : 0, std::memory_order_release);
}

void KeyboardCounter::ClearCounts(bool reset_pressed_state) {
    for (auto& count : counts_) {
        std::atomic_ref(count).store(0, std::memory_order_relaxed);
    }
    std::atomic_ref(total_count_).store(0, std::memory_order_relaxed);
    if (reset_pressed_state) {
        std::lock_guard lock(pressed_gate_);
        pressed_.clear();
    }
}

void KeyboardCounter::RemoveDevice(std::intptr_t device_handle) {
    std::lock_guard lock(pressed_gate_);
    std::erase_if(pressed_, [device_handle](const PressedKey& key) {
        return key.device_handle == device_handle;
    });
}

void KeyboardCounter::ReleaseKey(KeyId key_id) {
    std::lock_guard lock(pressed_gate_);
    std::erase_if(pressed_, [key_id](const PressedKey& key) { return key.key_id == key_id; });
}

KeyboardSnapshot KeyboardCounter::GetSnapshot() const {
    KeyboardSnapshot snapshot;
    for (int index = 0; index < kKeyCount; ++index) {
        snapshot.counts[static_cast<std::size_t>(index)] =
            std::atomic_ref<const std::int64_t>(counts_[static_cast<std::size_t>(index)]).load(std::memory_order_relaxed);
    }
    snapshot.total_count = std::atomic_ref<const std::int64_t>(total_count_).load(std::memory_order_relaxed);
    snapshot.is_paused = IsPaused();
    {
        std::lock_guard lock(pressed_gate_);
        for (const auto& key : pressed_) {
            snapshot.pressed_keys.insert(ToIndex(key.key_id));
        }
    }
    return snapshot;
}

}  // namespace keystats
